### ISSUE TITLE
pattern-not and metavariable regex not working together

### ISSUE THREAD

**Issue by Uzair614** [semgrep/semgrep#2368]
Here is the code that I want to run semgrep on
```
row, _ := db.ExecContext(ctx, "SELECT number")
row, _ := db.ExecContext(ctx, "SELECT number WHERE tenantId = 1")
```

^^ I want to catch lines WITHOUT the tenantId WHERE clause. So in the above case, line 1 should be flagged but line 2 should be fine.

The yaml config
```rules:
- id: invalid-query-detected
  languages: [go]
  message: Invalid Query Detected
  severity: WARNING
  patterns:
  - pattern: $OBJ.ExecContext(...)
 
  # pattern with metavariable-regex picking up only the second line
  # - pattern: $OBJ.ExecContext($CTX, $RAW_QUERY)

  # pattern-not with a static string argument works...
  # - pattern-not: $OBJ.ExecContext($CTX, "SELECT number WHERE tenantId = 1")

  # but pattern-not with a metavariable regex does not work
  - pattern-not: $OBJ.ExecContext($CTX, $RAW_QUERY)
  - metavariable-regex:
      metavariable: $RAW_QUERY
      regex: ".*tenantId.*"
```

(Note: the regex is a simple search for "tenantId" in the string for illustrative purposes)

A bit of rudimentary debugging on my side indicated that the reason no matches are found with the above config, is because t**he pattern and pattern-not match the same lines. The ideal case would be that the `pattern` matches both lines, and the `pattern-not` removes the correct (second) line.**

The metavariable-regex itself seems fine, because if instead I use only the single pattern
`
pattern: $OBJ.ExecContext($CTX, $RAW_QUERY)
`

then only the second line gets matched, signalling that the regex itself is fine, and that the problem lies with the interaction of pattern-not and metavariable regex.

https://semgrep.dev/s/KldY

**Comment by tvfmmfvt** [comment]
Not claiming but will use as exercise when bandwidth thx. Tight precise.

**Comment by colleend** [comment]
Thanks for opening such a well-written and detailed issue! 

It does appear that the metavariable-regex and pattern-not combination results in the pattern-not matching both the first and second lines. This is indeed a bug that we will look into (most likely a python wrapper issue). 

In the meantime, I've written something using the PCRE regex negative lookahead parameter (?!) in order to perform an operation similar to a `NOT`. This way, you can match the first line and not the second using pattern and metavariable regex:

https://semgrep.dev/s/Bk29/

Hope that helps, and we will definitely work to fix this issue.

**Comment by Uzair614** [comment]
Apologies for the late reply. Your workaround seems good, thanks.

**Comment by mschwager** [comment]
@brendongo @nbrahms do you know the status of this issue? I see it was moved into "To do... later" then moved back into "To do."

**Comment by nbrahms** [comment]
@Uzair614 what's the priority of this issue to you? It seems like you have a workaround now, so we'll put it at P2 for now. Please let us know if you want us to advance the priority of this. 🙇

**Comment by dkasak** [comment]
An alternative, and perhaps an easier solution, could be to add `metavariable-not-regex` and `metavariable-pattern-not` as analogues of `pattern-not-regex` and `pattern-not`. #3246 looks related.

**Comment by ceefour** [comment]
Agree with @dkasak ! Please add metavariable-not-regex and metavariable-pattern-not

**Comment by lfama** [comment]
Hi  @colleend ! 
Is there any update on this issue? Are there any plans to get it fixed?

**Comment by IagoAbal** [comment]
This has been fixed already! I think by #6486. See https://semgrep.dev/playground/s/XX0e.

You can now write:
```yaml
- pattern-not:
    patterns:
      - pattern: $OBJ.ExecContext($CTX, $RAW_QUERY)
      - metavariable-regex:
          metavariable: $RAW_QUERY
          regex: .*tenantId.*
```
